### PR TITLE
[Banners] Backport the fixes / changes from the banners repo into the CMS

### PR DESCRIPTION
**Pull request by zero-24**
For the changes see:
https://github.com/joomla-extensions/banners/pull/1
https://github.com/joomla-extensions/banners/pull/2
https://github.com/zero-24/banners/pull/1
https://github.com/zero-24/banners/pull/2
https://github.com/joomla/joomla-cms/pull/7009
https://github.com/joomla/joomla-cms/pull/6978
https://github.com/joomla/joomla-cms/pull/6946
### How to test

Please test be careful that com_banners and mod_banners still works as expected.

### CODE REVIEW

**Comment by smz** [comment]
@zero-24 Sorry, I may be thick but I don't understand how we should test this:
- Apply patch to an existing installation and test

or...
- Apply patch, perform installation anew, and **then** test

**Comment by zero-24** [comment]
@smz the first one as we don't finaly remove banners this only backports the fixes. But i guess i should remove the most changes.

**Comment by zero-24** [comment]
Just synced to staging and get some changes to the doc block out of this PR. Let me know if i should also remove the CS fixes.

**Comment by smz** [comment]
Tested as far as I could and found no issue...

Tested functionalities:
- Banners (Image type only, not custom)
- Categories
- Clients
- Tracks
- Banners versioning
- Banners and Categories batch operations
- Multilingual functionality
- **Untested with Postgress**

The only *anomaly* I found is that categories Language doesn't seems to have any effect: it is the Banners language that determine for which languages a particular banner is displayed...

**Comment by jwaisner** [comment]
@test 

After applying PR, no changes to operation of com_banners and mod_banners. Test successful.

Note: Not Tested on Postgress<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7337">issues.joomla.org/joomla-cms/7337</a>.</sub>

**Comment by smz** [comment]
It seems that the PLT decision to **not** uncouple extensions has been reverted (see: https://groups.google.com/d/msg/joomla-dev-cms/wLORxHBDSI0/iQxroGhCfBsJ): shouldn't we close this and re-open #7008 ?

**Comment by zero-24** [comment]
Based on the comment by @wilsonge in the Google Group and the tests I'm RTC'ing<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7337">issues.joomla.org/joomla-cms/7337</a>.</sub>

**Comment by zero-24** [comment]
@Kubik-Rubik I have just fixed the conflicts.

File: administrator/components/com_banners/models/banner.php [hunk 1]
```diff
@@ -478,8 +478,8 @@ protected function prepareTable($table)
 		else
 		{
 			// Set the values
-			$table->modified	= $date->toSql();
-			$table->modified_by	= $user->get('id');
+			$table->modified   = $date->toSql();
+			$table->created_by = $user->id;
```
Comment (wilsonge) [inline administrator/components/com_banners/models/banner.php]: What's going on with the change from modified_by to created_by here. I don't see that change in the banners component repo and it seems wrong as well

File: administrator/components/com_banners/sql/uninstall.sqlsrv.sql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+DELETE FROM [#__content_types] WHERE [type_alias] IN ('com_banners.banner', 'com_banners.category', 'com_banners.client');
+
+DROP TABLE IF EXISTS `#__banners`;
```
Comment (mbabker) [inline administrator/components/com_banners/sql/uninstall.sqlsrv.sql]: The quoting is wrong for the SQL Server engine.

File: administrator/components/com_banners/sql/uninstall.postgresql.sql [hunk 3]
```diff
@@ -0,0 +1,8 @@
+DELETE FROM "#__content_types" WHERE "type_alias" IN ('com_banners.banner', 'com_banners.category', 'com_banners.client');
+
+DROP TABLE IF EXISTS `#__banners`;
```
Comment (mbabker) [inline administrator/components/com_banners/sql/uninstall.postgresql.sql]: The quoting is wrong for the PostgreSQL engine.

File: administrator/components/com_banners/models/banner.php (hunk 1 above)
Comment (wilsonge) [inline administrator/components/com_banners/models/banner.php]: OK sorry I can see it in a code style commit you made. I think that's an error - can you confirm please

**Comment by wilsonge** [comment]
We also need to fix the conflicts again please (sorry)

File: administrator/components/com_banners/models/banner.php (hunk 1 above)
Comment (zero-24) [inline administrator/components/com_banners/models/banner.php]: @wilsonge see: https://github.com/joomla/joomla-cms/pull/7009

**Comment by zero-24** [comment]
@wilsonge just fixed.

**Comment by zero-24** [comment]
@mbabker I have also reverted all SQL Changes.

File: administrator/components/com_banners/views/client/tmpl/edit.php [hunk 4]
```diff
@@ -13,17 +13,19 @@
...
 		if (task == "client.cancel" || document.formvalidator.isValid(document.getElementById("client-form")))
 		{
 			Joomla.submitform(task, document.getElementById("client-form"));
 		}
-	};'
-);
+	};
+');
 ?>
+<script type="text/javascript">
```
Comment (wilsonge) [inline administrator/components/com_banners/views/client/tmpl/edit.php]: we need to remove the empty script tags here and in the file below

Comment (zero-24) [inline administrator/components/com_banners/views/client/tmpl/edit.php]: Fixed @wilsonge

**Comment by zero-24** [comment]
Maybe somone can have a look here. Nothing critical will be changed ;)

**Comment by dgt41** [comment]
I have tested this item :white_check_mark: successfully on dbd6c6f03b7f50610e6ec39b016a6b8968b0f26e<br /><br />Inception mode = ON <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7337">issues.joomla.org/joomla-cms/7337</a>.</sub>

**Comment by dgt41** [comment]
To test this go to /administrator/index.php?option=com_banners&view=tracks click on the button export and then test if the buttons Export and cancel are closing the modal<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7337">issues.joomla.org/joomla-cms/7337</a>.</sub>

**Comment by zero-24** [comment]
Thanks :smile:

**Pull request merged**